File: ScaleformUI_Csharp/BreadcrumbsHandler.cs
namespace ScaleformUI
{
    internal static class BreadcrumbsHandler
    {
        private static readonly List<UIMenu> breadcrumbs = new List<UIMenu>();
        internal static int Count => breadcrumbs.Count;
        internal static int CurrentDepth => breadcrumbs.Count - 1;
        internal static UIMenu PreviousMenu => breadcrumbs[CurrentDepth - 1];
        internal static bool SwitchInProgress = false;

        internal static void Forward(UIMenu menu)
        {
            breadcrumbs.Add(menu);
        }

        internal static void Clear()
        {
            breadcrumbs.Clear();
        }

        internal static void Backwards()
        {
            breadcrumbs.RemoveAt(CurrentDepth);
        }
    }
}

File: ScaleformUI_Lua/src/Menus/BreadcrumbsHandler.lua
BreadcrumbsHandler = setmetatable({
    breadcrumbs = {},
    SwitchInProgress = false
}, BreadcrumbsHandler)

function BreadcrumbsHandler:Count()
    return #self.breadcrumbs
end

function BreadcrumbsHandler:CurrentDepth()
    return #self.breadcrumbs -- needed? lua handles arrays from 1 not 0.. so..who knows..
end

function BreadcrumbsHandler:PreviousMenu()
    return self.breadcrumbs[#self.breadcrumbs - 1]
end

function BreadcrumbsHandler:Forward(menu, data)
    table.insert(self.breadcrumbs, { menu = menu, data = data })
end

function BreadcrumbsHandler:Clear()
    self.breadcrumbs = {}
end

function BreadcrumbsHandler:Backwards()
    table.remove(self.breadcrumbs, #self.breadcrumbs)
end
